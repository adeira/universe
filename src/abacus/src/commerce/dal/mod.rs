pub(in crate::commerce) mod products;

#[cfg(test)]
mod products_tests;
