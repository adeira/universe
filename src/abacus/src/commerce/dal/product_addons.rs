use crate::arango::{resolve_aql_vector, ConnectionPool};
use crate::commerce::model::product_addons::ProductAddon;
use crate::locale::SupportedLocale;

/// TODO(004) - integration tests
pub(in crate::commerce) async fn search_product_addons(
    pool: &ConnectionPool,
    client_locale: &SupportedLocale,
) -> anyhow::Result<Vec<Option<ProductAddon>>> {
    resolve_aql_vector(
        &pool,
        r#"
            FOR product_addon IN product_addons
              LET t = FIRST(
                FOR t IN product_addon.translations
                  FILTER t.name != null AND t.locale == @client_locale
                  RETURN t
              )
              SORT t.name ASC
              RETURN MERGE(
                product_addon,
                { name: t.name }
              )
        "#,
        hashmap_json![
            "client_locale" => client_locale
        ],
    )
    .await
}
