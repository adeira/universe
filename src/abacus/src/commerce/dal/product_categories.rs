use crate::arangodb::{resolve_aql_vector, ConnectionPool};
use crate::commerce::model::product_categories::ProductCategory;
use crate::locale::SupportedLocale;

pub(in crate::commerce) async fn search_all_product_categories(
    pool: &ConnectionPool,
    client_locale: &SupportedLocale,
) -> anyhow::Result<Vec<Option<ProductCategory>>> {
    resolve_aql_vector(
        &pool,
        r#"
            FOR product_category IN product_categories
              LET t = FIRST(
                FOR t IN product_category.translations
                  FILTER t.name != null AND t.locale == @client_locale
                  RETURN t
              )
              RETURN MERGE(
                product_category,
                { name: t.name }
              )
        "#,
        hashmap_json![
            "client_locale" => client_locale,
        ],
    )
    .await
}

pub(in crate::commerce) async fn get_product_categories_by_ids(
    pool: &ConnectionPool,
    client_locale: &SupportedLocale,
    product_category_ids: &[String],
) -> anyhow::Result<Vec<Option<ProductCategory>>> {
    resolve_aql_vector(
        &pool,
        r#"
            FOR product_category IN DOCUMENT(product_categories, @product_category_ids)
              LET t = FIRST(
                FOR t IN product_category.translations
                  FILTER t.name != null AND t.locale == @client_locale
                  RETURN t
              )
              RETURN MERGE(
                product_category,
                { name: t.name }
              )
        "#,
        hashmap_json![
            "client_locale" => client_locale,
            "product_category_ids" => product_category_ids,
        ],
    )
    .await
}
