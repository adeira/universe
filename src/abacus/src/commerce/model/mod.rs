pub(in crate::commerce) mod errors;
pub(in crate::commerce) mod products;
