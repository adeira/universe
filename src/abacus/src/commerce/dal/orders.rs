use crate::arango::{resolve_aql_vector, ConnectionPool};
use crate::commerce::model::product_addons::ProductAddon;
use crate::locale::SupportedLocale;

/// The following function creates a payment that is awaiting payment. It should not be processed
/// yet until we actually receive the money (confirmation from Stripe.com).
///
/// TODO(004) - integration tests
pub(in crate::commerce) async fn create_unpaid_order(
    pool: &ConnectionPool,
    client_locale: &SupportedLocale,
) -> anyhow::Result<Vec<Option<ProductAddon>>> {
    resolve_aql_vector(
        pool,
        r#"
            TODO
        "#,
        hashmap_json![
            "client_locale" => client_locale
        ],
    )
    .await
}
