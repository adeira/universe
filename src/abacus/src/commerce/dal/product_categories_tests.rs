use crate::arangodb::{cleanup_test_database, prepare_empty_test_database};
use crate::commerce::dal::product_categories::{
    get_product_categories_by_ids, search_all_product_categories,
};
use crate::locale::SupportedLocale;

// (cd src/abacus && cargo t -- --ignored commerce::dal::product_categories_tests)

#[ignore]
#[tokio::test]
async fn search_all_product_categories_test() {
    let db_name = "search_all_product_categories_test";
    let pool = prepare_empty_test_database(&db_name).await;

    // 1) try to find all product categories in ENGLISH
    let found_categories_en = search_all_product_categories(&pool, &SupportedLocale::EnUS)
        .await
        .unwrap();

    assert_eq!(found_categories_en.len(), 2);
    insta::assert_debug_snapshot!(found_categories_en);

    // 2) try to find all product categories in SPANISH
    let found_categories_en = search_all_product_categories(&pool, &SupportedLocale::EsMX)
        .await
        .unwrap();

    assert_eq!(found_categories_en.len(), 2);
    insta::assert_debug_snapshot!(found_categories_en);

    cleanup_test_database(&db_name).await;
}

#[ignore]
#[tokio::test]
async fn get_product_categories_by_ids_test() {
    let db_name = "get_product_categories_by_ids_test";
    let pool = prepare_empty_test_database(&db_name).await;

    // 1) try to find product categories in ENGLISH by IDs
    let found_categories_en = get_product_categories_by_ids(
        &pool,
        &SupportedLocale::EnUS,
        &[
            "product_categories/coffee".to_string(), // added by DB migration
            "product_categories/tea".to_string(),    // added by DB migration
        ],
    )
    .await
    .unwrap();

    assert_eq!(found_categories_en.len(), 2);
    insta::assert_debug_snapshot!(found_categories_en);

    // 2) try to find one product category in SPANISH using DB key
    let found_categories_en = get_product_categories_by_ids(
        &pool,
        &SupportedLocale::EsMX,
        &[
            "coffee".to_string(), // notice it's not a DB `_id` but DB `_key`
        ],
    )
    .await
    .unwrap();

    assert_eq!(found_categories_en.len(), 1);
    insta::assert_debug_snapshot!(found_categories_en);

    cleanup_test_database(&db_name).await;
}
