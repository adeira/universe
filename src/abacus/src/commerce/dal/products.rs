use crate::arangodb::{resolve_aql, resolve_aql_vector, ConnectionPool};
use crate::commerce::model::products::{
    PriceSortDirection, Product, ProductMultilingualInput, ProductMultilingualInputVisibility,
};
use crate::images::Image;
use crate::locale::SupportedLocale;
use serde_json::json;

/// Takes care of creating the product inside ArangoDB.
pub(in crate::commerce) async fn create_product(
    pool: &ConnectionPool,
    product_multilingual_input: &ProductMultilingualInput,
    images: &[Image],
) -> anyhow::Result<Product> {
    // TODO: dynamic `unit_label`
    // TODO: https://www.arangodb.com/docs/stable/aql/extending.html (for merging translations)
    resolve_aql(
        &pool,
        r#"
            LET unit_label_translated = DOCUMENT("product_units/piece")[@eshop_locale]

            INSERT {
              images: @product_images,
              unit_label: "product_units/piece",
              is_published: false,
              visibility: @product_visibility,
              created: DATE_ISO8601(DATE_NOW()),
              updated: DATE_ISO8601(DATE_NOW()),
              price: {
                unit_amount: @product_price_unit_amount,
                unit_amount_currency: "MXN",
              },
              translations: @translations
            } INTO products
            LET product = NEW

            LET t = FIRST(
              FOR t IN product.translations
                FILTER t.name != null AND t.locale == @eshop_locale
                RETURN t
            )

            RETURN MERGE(
              product,
              { unit_label: unit_label_translated },
              { name: t.name, description: t.description }
            )
        "#,
        hashmap_json![
            "eshop_locale" => SupportedLocale::EnUS, // TODO
            "product_images" => images,
            "product_visibility" => product_multilingual_input.visibility,
            "product_price_unit_amount" => product_multilingual_input.price.unit_amount,
            "translations" => product_multilingual_input.translations,
        ],
    )
    .await
}

/// TODO(004) - integration tests
pub(in crate::commerce) async fn update_product(
    pool: &ConnectionPool,
    product_key: &str,
    product_revision: &str,
    product_multilingual_input: &ProductMultilingualInput,
    images: &[Image],
) -> anyhow::Result<Product> {
    // TODO: https://www.arangodb.com/docs/stable/aql/extending.html (for merging translations)
    resolve_aql(
        &pool,
        r#"
            LET unit_label_translated = DOCUMENT("product_units/piece")[@eshop_locale]

            UPDATE {
              _key: @product_key,
              _rev: @product_rev,
              images: @product_images,
              visibility: @product_visibility,
              updated: DATE_ISO8601(DATE_NOW()),
              price: {
                unit_amount: @product_price_unit_amount,
                unit_amount_currency: "MXN",
              },
              translations: @translations
            } IN products OPTIONS { ignoreRevs: false }
            LET product = NEW

            LET t = FIRST(
              FOR t IN product.translations
                FILTER t.name != null AND t.locale == @eshop_locale
                RETURN t
            )

            RETURN MERGE(
              product,
              { unit_label: unit_label_translated },
              { name: t.name, description: t.description }
            )
        "#,
        hashmap_json![
            "eshop_locale" => SupportedLocale::EnUS, // TODO
            "product_key" => product_key,
            "product_rev" => product_revision,
            "product_images" => images,
            "product_visibility" => product_multilingual_input.visibility,
            "product_price_unit_amount" => product_multilingual_input.price.unit_amount,
            "translations" => product_multilingual_input.translations,
        ],
    )
    .await
}

/// TODO(004) - integration tests
pub(in crate::commerce) async fn publish_product(
    pool: &ConnectionPool,
    product_key: &str,
) -> anyhow::Result<Product> {
    // TODO: https://www.arangodb.com/docs/stable/aql/extending.html (for merging translations)
    resolve_aql(
        &pool,
        r#"
            LET unit_label_translated = DOCUMENT("product_units/piece")[@eshop_locale]

            UPDATE {
              _key: @product_key,
              is_published: true,
            } IN products
            LET product = NEW

            LET t = FIRST(
              FOR t IN product.translations
                FILTER t.name != null AND t.locale == @eshop_locale
                RETURN t
            )

            RETURN MERGE(
              product,
              { unit_label: unit_label_translated },
              { name: t.name, description: t.description }
            )
        "#,
        hashmap_json![
            "eshop_locale" => SupportedLocale::EnUS, // TODO
            "product_key" => product_key,
        ],
    )
    .await
}

/// TODO(004) - integration tests
pub(in crate::commerce) async fn unpublish_product(
    pool: &ConnectionPool,
    product_key: &str,
) -> anyhow::Result<Product> {
    // TODO: https://www.arangodb.com/docs/stable/aql/extending.html (for merging translations)
    resolve_aql(
        &pool,
        r#"
            LET unit_label_translated = DOCUMENT("product_units/piece")[@eshop_locale]

            UPDATE {
              _key: @product_key,
              is_published: false,
            } IN products
            LET product = NEW

            LET t = FIRST(
              FOR t IN product.translations
                FILTER t.name != null AND t.locale == @eshop_locale
                RETURN t
            )

            RETURN MERGE(
              product,
              { unit_label: unit_label_translated },
              { name: t.name, description: t.description }
            )
        "#,
        hashmap_json![
            "eshop_locale" => SupportedLocale::EnUS, // TODO
            "product_key" => product_key,
        ],
    )
    .await
}

/// Returns a single product (or error).
pub(in crate::commerce) async fn get_product_by_key(
    pool: &ConnectionPool,
    client_locale: &SupportedLocale,
    product_key: &str,
    product_published_only: &bool,
) -> anyhow::Result<Product> {
    let product_vector = get_products_by_keys(
        &pool,
        &client_locale,
        &[product_key.to_string()],
        &product_published_only,
    )
    .await?;

    match product_vector.first() {
        Some(product) => Ok(product.to_owned()),
        None => anyhow::bail!("database didn't return any product"),
    }
}

pub(in crate::commerce) async fn get_products_by_keys(
    pool: &ConnectionPool,
    client_locale: &SupportedLocale,
    product_keys: &[String],
    product_published_only: &bool,
) -> anyhow::Result<Vec<Product>> {
    // TODO: https://www.arangodb.com/docs/stable/aql/extending.html (for merging translations)
    resolve_aql_vector(
        &pool,
        r#"
            LET products = DOCUMENT(products, @product_keys)
            FOR product IN products
              FILTER @product_published_only ? product.is_published == @product_published_only : true

              LET t = FIRST(
                FOR t IN product.translations
                  FILTER t.name != null AND t.locale == @eshop_locale
                  RETURN t
              )

              RETURN MERGE(
                product,
                { unit_label: DOCUMENT(product.unit_label)[@eshop_locale] },
                { name: t.name, description: t.description }
              )
        "#,
        hashmap_json![
            "eshop_locale" => client_locale,
            "product_keys" => product_keys,
            "product_published_only" => product_published_only,
        ]
    ).await
}

/// Performs search of products based on the specified criteria and returns products with merged
/// translations based on the eshop language.
///
/// Optionally, you can specify a search term. In this case, it performs the same search except it
/// performs additional fulltext search and additionally sorts the results by relevance.
///
/// TODO(004) - integration tests
pub(in crate::commerce) async fn search_products(
    pool: &ConnectionPool,
    client_locale: &SupportedLocale,
    price_sort_direction: &PriceSortDirection,
    search_term: &Option<String>,
    search_all: &bool,
    visibility: &Option<ProductMultilingualInputVisibility>,
) -> anyhow::Result<Vec<Option<Product>>> {
    let sort_direction = match price_sort_direction {
        PriceSortDirection::LowToHigh => "ASC",
        PriceSortDirection::HighToLow => "DESC",
    };

    let visibility = match visibility {
        Some(visibility) => json!(visibility),
        None => json!(null),
    };

    match search_term {
        Some(search_term) => {
            // TODO: https://www.arangodb.com/docs/stable/aql/extending.html (for merging translations)
            resolve_aql_vector(
                &pool,
                // This should be almost identical with the query below except it uses fulltext
                // search across all supported languages and it additionally sorts the results by
                // relevance:
                r#"
                    FOR product IN search_products
                      SEARCH BOOST(NGRAM_MATCH(product.translations.name, @search_term, 0.7, "bigram"), 1.1)
                          OR BOOST(NGRAM_MATCH(product.translations.description, @search_term, 0.7, "bigram"), 1.0)
                      FILTER @search_all == true ? true : (product.is_published IN [true])
                      FILTER @visibility == null ? true : (@visibility IN product.visibility)
                      SORT BM25(product) DESC
                      SORT product.price.unit_amount @price_sort_direction

                      LET t = FIRST(
                        FOR t IN product.translations
                          FILTER t.name != null AND t.locale == @eshop_locale
                          RETURN t
                      )

                      RETURN MERGE(
                        product,
                        { unit_label: DOCUMENT(product.unit_label)[@eshop_locale] },
                        { name: t.name, description: t.description }
                      )
                "#,
                hashmap_json![
                    "eshop_locale" => client_locale,
                    "search_term" => search_term,
                    "search_all" => search_all,
                    "price_sort_direction" => sort_direction,
                    "visibility" => visibility,
                ]
            ).await
        }
        None => {
            // TODO: https://www.arangodb.com/docs/stable/aql/extending.html (for merging translations)
            resolve_aql_vector(
                &pool,
                r#"
                    FOR product IN search_products
                      FILTER @search_all == true ? true : (product.is_published IN [true])
                      FILTER @visibility == null ? true : (@visibility IN product.visibility)
                      SORT product.price.unit_amount @price_sort_direction

                      LET t = FIRST(
                        FOR t IN product.translations
                          FILTER t.name != null AND t.locale == @eshop_locale
                          RETURN t
                      )

                      RETURN MERGE(
                        product,
                        { unit_label: DOCUMENT(product.unit_label)[@eshop_locale] },
                        { name: t.name, description: t.description }
                      )
                "#,
                hashmap_json![
                    "eshop_locale" => client_locale,
                    "search_all" => search_all,
                    "price_sort_direction" => sort_direction,
                    "visibility" => visibility,
                ],
            )
            .await
        }
    }
}

/// Important note: product should be moved into the archive before deleting it!
pub(in crate::commerce) async fn delete_product(
    pool: &ConnectionPool,
    product_key: &str,
) -> anyhow::Result<Product> {
    resolve_aql(
        &pool,
        r#"
            LET unit_label_translated = DOCUMENT("product_units/piece")[@eshop_locale]
            LET product = DOCUMENT(products, @product_key)
            REMOVE product IN products

            LET t = FIRST(
              FOR t IN product.translations
                FILTER t.name != null AND t.locale == @eshop_locale
                RETURN t
            )

            RETURN MERGE(
              product,
              { unit_label: unit_label_translated },
              { name: t.name, description: t.description }
            )
        "#,
        hashmap_json![
            "eshop_locale" => SupportedLocale::EnUS, // TODO
            "product_key" => product_key,
        ],
    )
    .await
}
