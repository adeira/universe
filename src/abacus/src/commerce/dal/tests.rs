use crate::arangodb::{cleanup_test_database, prepare_empty_test_database};
use crate::commerce::api::PriceSortDirection;
use crate::commerce::dal::products::{
    create_product, delete_product, get_product_by_key, search_products,
};
use crate::commerce::model::products::{
    ProductMultilingualInput, ProductMultilingualInputTranslations, SupportedLocale,
};

#[ignore]
#[tokio::test]
async fn create_product_english_to_english_test() {
    let db_name = "create_product_english_to_english_test";
    let pool = prepare_empty_test_database(&db_name).await;

    // 1) create a product with english name and description
    let created_product = create_product(
        &pool,
        &ProductMultilingualInput {
            translations: vec![ProductMultilingualInputTranslations {
                locale: SupportedLocale::EnUS,
                name: "Product name in english".to_string(),
                description_slate: None,
            }],
            ..Default::default()
        },
        &[],
    )
    .await
    .unwrap();

    // 2) try to find the newly created product
    let found_product = get_product_by_key(
        &pool,
        &SupportedLocale::EnUS,
        &created_product.key(),
        &false, // the product should be unpublished at this point (until manually activated)
    )
    .await
    .unwrap();
    assert_eq!(found_product.name(), "Product name in english".to_string());

    cleanup_test_database(&db_name).await;
}

#[ignore]
#[tokio::test]
async fn create_product_all_languages_test() {
    let db_name = "create_product_all_languages_test";
    let pool = prepare_empty_test_database(&db_name).await;

    // 1) create a product with english AND spanish name and description
    let created_product = create_product(
        &pool,
        &ProductMultilingualInput {
            translations: vec![
                ProductMultilingualInputTranslations {
                    locale: SupportedLocale::EnUS,
                    name: "Product name in english".to_string(),
                    description_slate: None,
                },
                ProductMultilingualInputTranslations {
                    locale: SupportedLocale::EsMX,
                    name: "Product name in SPANISH".to_string(),
                    description_slate: None,
                },
            ],
            ..Default::default()
        },
        &[],
    )
    .await
    .unwrap();

    // 2) try to find the newly created product in both client locales
    let found_en_product = get_product_by_key(
        &pool,
        &SupportedLocale::EnUS,
        &created_product.key(),
        &false, // the product should be unpublished at this point (until manually activated)
    )
    .await
    .unwrap();
    assert_eq!(
        found_en_product.name(),
        "Product name in english".to_string()
    );

    let found_es_product = get_product_by_key(
        &pool,
        &SupportedLocale::EsMX,
        &created_product.key(),
        &false, // the product should be unpublished at this point (until manually activated)
    )
    .await
    .unwrap();
    assert_eq!(
        found_es_product.name(),
        "Product name in SPANISH".to_string()
    );

    cleanup_test_database(&db_name).await;
}

#[ignore]
#[tokio::test]
async fn search_products_test() {
    let db_name = "search_products_test";
    let pool = prepare_empty_test_database(&db_name).await;

    // 1) create a product to be later searched
    create_product(
        &pool,
        &ProductMultilingualInput {
            translations: vec![ProductMultilingualInputTranslations {
                locale: SupportedLocale::EnUS,
                name: "Product name in english".to_string(),
                description_slate: None,
            }],
            ..Default::default()
        },
        &[],
    )
    .await
    .unwrap();

    // TODO: This is super ugly (potentially flaky) but we need to wait for the search view
    //       to be ready. How to do it better?
    let five_seconds = std::time::Duration::from_secs(5);
    std::thread::sleep(five_seconds);

    // 2) try to search all products
    let searched_products = search_products(
        &pool,
        &SupportedLocale::EnUS,
        &PriceSortDirection::LowToHigh,
        &None, // search term (return all)
        &true, // search all (even the unpublished ones)
        &None, // visibility (everywhere)
    )
    .await
    .unwrap();

    insta::assert_debug_snapshot!(searched_products);

    cleanup_test_database(&db_name).await;
}

#[ignore]
#[tokio::test]
async fn search_products_fulltext_test() {
    let db_name = "search_products_fulltext_test";
    let pool = prepare_empty_test_database(&db_name).await;

    // 1) create a product to be later searched
    create_product(
        &pool,
        &ProductMultilingualInput {
            translations: vec![ProductMultilingualInputTranslations {
                locale: SupportedLocale::EnUS,
                name: "Product name in english".to_string(),
                description_slate: None,
            }],
            ..Default::default()
        },
        &[],
    )
    .await
    .unwrap();

    // TODO: This is super ugly (potentially flaky) but we need to wait for the search view
    //       to be ready. How to do it better?
    let five_seconds = std::time::Duration::from_secs(5);
    std::thread::sleep(five_seconds);

    // 2) try to search the product
    let searched_products = search_products(
        &pool,
        &SupportedLocale::EnUS,
        &PriceSortDirection::LowToHigh,
        &Some(String::from("in english")), // search term
        &true,                             // search all (even the unpublished ones)
        &None,                             // visibility (everywhere)
    )
    .await
    .unwrap();

    insta::assert_debug_snapshot!(searched_products);

    cleanup_test_database(&db_name).await;
}

#[ignore]
#[tokio::test]
async fn delete_product_test() {
    let db_name = "delete_product_test";
    let pool = prepare_empty_test_database(&db_name).await;

    // 1) create a product to be later deleted
    let created_product = create_product(
        &pool,
        &ProductMultilingualInput {
            translations: vec![ProductMultilingualInputTranslations {
                locale: SupportedLocale::EnUS,
                name: "Product name in english".to_string(),
                description_slate: None,
            }],
            ..Default::default()
        },
        &[],
    )
    .await
    .unwrap();

    // 2) try to delete the newly created product
    let deleted_product = delete_product(&pool, &created_product.key()).await.unwrap();
    assert_eq!(
        deleted_product.name(),
        "Product name in english".to_string()
    );

    cleanup_test_database(&db_name).await;
}
