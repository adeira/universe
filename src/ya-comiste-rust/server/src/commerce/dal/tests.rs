use crate::arangodb::{cleanup_test_database, prepare_empty_test_database};
use crate::commerce::dal::products::{create_product, delete_product, get_product_by_key};
use crate::commerce::model::products::{
    ProductMultilingualInput, ProductMultilingualInputTranslations, SupportedLocale,
};

#[ignore]
#[tokio::test]
async fn create_product_english_to_english_test() {
    let db_name = "create_product_english_to_english_test";
    let pool = prepare_empty_test_database(&db_name).await;

    // 1) create a product with english name and description
    let created_product = create_product(
        &pool,
        &ProductMultilingualInput {
            translations: vec![ProductMultilingualInputTranslations {
                locale: SupportedLocale::EnUS,
                name: Some("Product name in english".to_string()),
                description: Some("Product description in english".to_string()),
            }],
            ..Default::default()
        },
        &vec![],
    )
    .await
    .unwrap();

    // 2) try to find the newly created product
    let found_product = get_product_by_key(
        &pool,
        &SupportedLocale::EnUS,
        &created_product.key(),
        &false, // the product should be inactive at this point (until manually activated)
    )
    .await
    .unwrap();
    assert_eq!(
        found_product.name(),
        Some("Product name in english".to_string())
    );

    cleanup_test_database(&db_name).await;
}

#[ignore]
#[tokio::test]
async fn create_product_spanish_to_english_test() {
    let db_name = "create_product_spanish_to_english_test";
    let pool = prepare_empty_test_database(&db_name).await;

    // 1) create a product with spanish name and description
    let created_product = create_product(
        &pool,
        &ProductMultilingualInput {
            translations: vec![ProductMultilingualInputTranslations {
                locale: SupportedLocale::EsMX,
                name: Some("Product name in SPANISH".to_string()),
                description: Some("Product description in SPANISH".to_string()),
            }],
            ..Default::default()
        },
        &vec![],
    )
    .await
    .unwrap();

    // 2) try to find the newly created product - please note that we are trying to fetch the
    //    english version even though it's not available and the result should fallback to the
    //    first available language which is spanish in this case
    let found_product = get_product_by_key(
        &pool,
        &SupportedLocale::EnUS,
        &created_product.key(),
        &false, // the product should be inactive at this point (until manually activated)
    )
    .await
    .unwrap();
    assert_eq!(
        found_product.name(),
        Some("Product name in SPANISH".to_string())
    );

    cleanup_test_database(&db_name).await;
}

#[ignore]
#[tokio::test]
async fn create_product_all_languages_test() {
    let db_name = "create_product_all_languages_test";
    let pool = prepare_empty_test_database(&db_name).await;

    // 1) create a product with english AND spanish name and description
    let created_product = create_product(
        &pool,
        &ProductMultilingualInput {
            translations: vec![
                ProductMultilingualInputTranslations {
                    locale: SupportedLocale::EnUS,
                    name: Some("Product name in english".to_string()),
                    description: Some("Product description in english".to_string()),
                },
                ProductMultilingualInputTranslations {
                    locale: SupportedLocale::EsMX,
                    name: Some("Product name in SPANISH".to_string()),
                    description: Some("Product description in SPANISH".to_string()),
                },
            ],
            ..Default::default()
        },
        &vec![],
    )
    .await
    .unwrap();

    // 2) try to find the newly created product in both client locales
    let found_en_product = get_product_by_key(
        &pool,
        &SupportedLocale::EnUS,
        &created_product.key(),
        &false, // the product should be inactive at this point (until manually activated)
    )
    .await
    .unwrap();
    assert_eq!(
        found_en_product.name(),
        Some("Product name in english".to_string())
    );

    let found_es_product = get_product_by_key(
        &pool,
        &SupportedLocale::EsMX,
        &created_product.key(),
        &false, // the product should be inactive at this point (until manually activated)
    )
    .await
    .unwrap();
    assert_eq!(
        found_es_product.name(),
        Some("Product name in SPANISH".to_string())
    );

    cleanup_test_database(&db_name).await;
}

#[ignore]
#[tokio::test]
async fn create_product_mixed_languages_test() {
    let db_name = "create_product_mixed_languages_test";
    let pool = prepare_empty_test_database(&db_name).await;

    // 1) create a product with english name AND spanish description (notice how it's mixed)
    let created_product = create_product(
        &pool,
        &ProductMultilingualInput {
            translations: vec![
                ProductMultilingualInputTranslations {
                    locale: SupportedLocale::EnUS,
                    name: Some("Product name in english".to_string()),
                    description: None,
                },
                ProductMultilingualInputTranslations {
                    locale: SupportedLocale::EsMX,
                    name: None,
                    description: Some("Product description in SPANISH".to_string()),
                },
            ],
            ..Default::default()
        },
        &vec![],
    )
    .await
    .unwrap();

    // 2) try to find the newly created product in both client locales
    let found_en_product = get_product_by_key(
        &pool,
        &SupportedLocale::EnUS,
        &created_product.key(),
        &false, // the product should be inactive at this point (until manually activated)
    )
    .await
    .unwrap();
    assert_eq!(
        found_en_product.name(),
        Some("Product name in english".to_string())
    );
    assert_eq!(found_en_product.description(), None);

    let found_es_product = get_product_by_key(
        &pool,
        &SupportedLocale::EsMX,
        &created_product.key(),
        &false, // the product should be inactive at this point (until manually activated)
    )
    .await
    .unwrap();
    assert_eq!(
        found_es_product.name(),
        Some("Product name in english".to_string()) // we fall back to EN because ES name is missing
    );
    assert_eq!(found_es_product.description(), None);

    cleanup_test_database(&db_name).await;
}

#[ignore]
#[tokio::test]
async fn delete_product_test() {
    let db_name = "delete_product_test";
    let pool = prepare_empty_test_database(&db_name).await;

    // 1) create a product to be later deleted
    let created_product = create_product(
        &pool,
        &ProductMultilingualInput {
            translations: vec![ProductMultilingualInputTranslations {
                locale: SupportedLocale::EnUS,
                name: Some("Product name in english".to_string()),
                description: Some("Product description in english".to_string()),
            }],
            ..Default::default()
        },
        &vec![],
    )
    .await
    .unwrap();

    // 2) try to delete the newly created product
    let deleted_product = delete_product(&pool, &created_product.key()).await.unwrap();
    assert_eq!(
        deleted_product.name(),
        Some("Product name in english".to_string())
    );

    cleanup_test_database(&db_name).await;
}
