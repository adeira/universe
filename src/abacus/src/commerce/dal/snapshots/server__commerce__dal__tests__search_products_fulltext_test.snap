---
source: server/src/commerce/dal/tests.rs
expression: searched_products
---
[
    Some(
        Product {
            name: "Product name in english",
            description: Some(
                "Product description in english",
            ),
            images: [],
            unit_label: "piece",
            is_published: false,
            visibility: [],
            price: ProductPrice {
                unit_amount: 0,
                unit_amount_currency: MXN,
            },
            translations: [
                ProductMultilingualTranslations {
                    locale: EnUS,
                    name: "Product name in english",
                    description: Some(
                        "Product description in english",
                    ),
                },
            ],
        },
    ),
]
